AI-AGENTS.md

```diff
@@ -0,0 +1,142 @@
+- Use `find_each()` or `in_batches()` for large dataset processing
+- Prefer database-level operations (`update_all`, `delete_all`) over Ruby loops for bulk changes
+- Use `exists?` instead of `present?` when checking for record existence
+
+### Migration Guidelines
+- Always include rollback logic in migrations
+- Use `add_index(..., algorithm: :concurrently)` for large tables in production
+- Never remove columns directly - deprecate first, then remove in subsequent release
+- Test migrations on production-sized datasets when possible
+
+### Query Optimization
+- Use `explain` to analyze query performance during development
+- Avoid `SELECT *` - specify needed columns explicitly
+- Use database indexes strategically, but avoid over-indexing
+- Consider using `counter_cache` for frequently accessed counts
+
+## Security Guidelines
+
+### XSS Prevention
+- Always use `{{}}` (escaped) instead of `{{{ }}}` (unescaped) in Ember templates
+- Sanitize user input using Discourse's built-in helpers (`sanitize`, `cook`)
+- Never directly insert user content into `innerHTML` or similar DOM methods
+- Use `@html` argument carefully and only with pre-sanitized content
+
+### CSRF Protection
+- All state-changing requests must use POST/PUT/DELETE, never GET
+- Ensure CSRF tokens are included in AJAX requests
+- Use Rails' `protect_from_forgery` in controllers handling sensitive operations
+
+### Input Sanitization
+- Validate and sanitize all user inputs on both client and server side
+- Use strong parameters in Rails controllers
+- Apply appropriate length limits and format validation
+- Never trust client-side validation alone
+
+### Authorization
+- Always use Guardian classes for authorization checks
+- Check permissions at both route and action levels
+- Implement proper scope limiting (users should only see their own data)
+- Use `can_see?` and `can_edit?` patterns consistently
+
+## Knowledge Sharing and Persistence
+
+- When asked to remember something, ALWAYS persist this information in a way that's accessible to ALL developers, not just in conversational memory
+- Document important information in appropriate files (comments, documentation, README, etc.) so other developers (human or AI) can access it
+- Information should be stored in a structured way that follows project conventions
+- NEVER keep crucial information only in conversational memory - this creates knowledge silos
+- If asked to implement something that won't be accessible to other users/developers in the repository, proactively highlight this issue
+- The goal is complete knowledge sharing between ALL developers (human and AI) without exceptions
+- When suggesting where to store information, recommend appropriate locations based on the type of information (code comments, documentation files, AI-AGENTS.md, etc.)
+- Inform the developer when you detect a change in this file and have successfully reloaded it.
```

Comment:
I haven't been able to make it do this consistently, it'd be helpful while we're iterating on this regularly, but it's probably fine if we ultimately remove it.